File: android/core/src/main/java/com/americangroupllc/drift/core/domain/LocationFuzzer.kt
package com.americangroupllc.drift.core.domain

import kotlinx.serialization.Serializable

@Serializable
data class FuzzedLocation(
    val zipPrefix3: String? = null,
    val countyFips: String? = null,
    val stateCode:  String? = null,
) {
    companion object {
        val EMPTY = FuzzedLocation()
    }
}

/** Pure-logic mirror of `LocationFuzzer.swift`. */
object LocationFuzzer {

    fun truncateZip(zip: String?): String? {
        if (zip.isNullOrBlank()) return null
        val digits = zip.filter { it.isDigit() }
        return if (digits.length >= 3) digits.take(3) else null
    }

    fun fuzz(zip5: String? = null, countyFips: String? = null, stateCode: String? = null): FuzzedLocation {
        val zip3 = truncateZip(zip5)
        val cf   = countyFips?.takeIf { it.length == 5 }
        val sc   = stateCode?.takeIf { it.length == 2 }?.uppercase()
        return FuzzedLocation(zip3, cf, sc)
    }

    fun validateCoordinate(lat: Double, lon: Double): Boolean =
        lat.isFinite() && lon.isFinite() &&
        lat in -90.0..90.0 && lon in -180.0..180.0
}

File: apps/DriftDate/shared/DriftCore/Sources/DriftCore/Domain/LocationFuzzer.swift
import Foundation

/// Pure-logic location fuzzing. Same input → same output. Never returns
/// the original lat/lon back to the caller.
///
/// `LocationFuzzer.fuzz(...)` is intentionally a value transformation: the
/// caller passes lat/lon, and the result struct does NOT contain coordinate
/// fields. The fuzzer is the canonical client-side path. The server's
/// `fuzz-location` Edge Function exists only as defence-in-depth.
public struct FuzzedLocation: Codable, Equatable, Sendable {
    public let zipPrefix3: String?
    public let countyFips: String?
    public let stateCode:  String?

    public init(zipPrefix3: String? = nil, countyFips: String? = nil, stateCode: String? = nil) {
        self.zipPrefix3 = zipPrefix3
        self.countyFips = countyFips
        self.stateCode  = stateCode
    }

    public static let empty = FuzzedLocation()
}

public enum LocationFuzzer {

    /// Truncate a ZIP-5 (or ZIP-9) string to its 3-char prefix.
    public static func truncateZip(_ zip: String?) -> String? {
        guard let zip else { return nil }
        let digits = zip.filter { $0.isNumber }
        guard digits.count >= 3 else { return nil }
        return String(digits.prefix(3))
    }

    /// Convenience: fuzz from already-resolved fields. Used when the
    /// caller already knows the ZIP and/or county and just wants a
    /// `FuzzedLocation` value with the truncation guarantees.
    public static func fuzz(
        zip5: String? = nil,
        countyFips: String? = nil,
        stateCode: String? = nil
    ) -> FuzzedLocation {
        let zip3 = truncateZip(zip5)
        let cf   = countyFips.flatMap { $0.count == 5 ? $0 : nil }
        let sc   = stateCode.flatMap { $0.count == 2 ? $0.uppercased() : nil }
        return FuzzedLocation(zipPrefix3: zip3, countyFips: cf, stateCode: sc)
    }

    /// Validate a lat/lon pair. Out-of-range or NaN inputs return nil so the
    /// caller can short-circuit any network call.
    public static func validateCoordinate(lat: Double, lon: Double) -> Bool {
        lat.isFinite && lon.isFinite &&
        lat >= -90  && lat <=  90  &&
        lon >= -180 && lon <= 180
    }
}
